## Homework #1

### Set up the environment

In [2]:
import pandas as pd
import numpy as np

### Question 1 Pandas version
What's the version of Pandas that you installed?

In [4]:
pd.__version__

'2.3.1'

The pandas version that I installed is '2.3.1'.

### Getting the data
For this homework, we'll use the Car Fuel Efficiency dataset. I am saving the data using the wget function for further review.

In [6]:
data = "https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv"
!wget -O cardata.csv $data

--2025-09-18 15:38:48--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.111.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 874188 (854K) [text/plain]
Saving to: ‘cardata.csv’

cardata.csv         100%[===================>] 853.70K  --.-KB/s    in 0.008s  

2025-09-18 15:38:48 (109 MB/s) - ‘cardata.csv’ saved [874188/874188]



In [7]:
df = pd.read_csv("cardata.csv")
df.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0         NaN     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

### Question 2. Records count
How many records are in the dataset?

In [11]:
df.shape[0]

9704

df.shape get the total rows and columns in tuple (x,y). We wants to know only how many records are there (which means the rows count) we just need to extract the first value of df.shape to get how many records. 9704 is the total record.

### Question 3. Fuel types
How many fuel types are presented in the dataset?

In [15]:
df.fuel_type.nunique()

2

With nunique, you can get the number of unique values directly. There are only 2 types of fuels presented in the dataset.

### Question 4. Missing values
How many columns in the dataset have missing values?

In [19]:
df.isna().any().sum()

np.int64(4)

In [20]:
df.columns[df.isna().any()]

Index(['num_cylinders', 'horsepower', 'acceleration', 'num_doors'], dtype='object')

There are four missing columns in the dataset.

### Question 5. Max fuel efficiency
What's the maximum fuel efficiency of cars from Asia?

In [21]:
df[df.origin == 'Asia']['fuel_efficiency_mpg'].max()

np.float64(23.759122836520497)

The maximum fuel efficiency of cars from Asia is 23.75.

### Question 6. Median value of horsepower
1. Find the median value of horsepower column in the dataset.
2. Next, calculate the most frequent value of the same horsepower column.
3. Use fillna method to fill the missing values in horsepower column with the most frequent value from the previous step.
4. Now, calculate the median value of horsepower once again.

In [27]:
df['horsepower'].value_counts().head(10)

horsepower
152.0    142
145.0    141
151.0    134
148.0    130
141.0    130
143.0    128
159.0    126
149.0    124
136.0    123
139.0    123
Name: count, dtype: int64

In [23]:
median_hp = df['horsepower'].median()
median_hp

np.float64(149.0)

In [24]:
frequent_hp = df['horsepower'].mode()[0]
frequent_hp

np.float64(152.0)

In [25]:
df['horsepower'].fillna(frequent_hp).median()

np.float64(152.0)

### Question 7. Sum of weights
1. Select all the cars from Asia
2. Select only columns vehicle_weight and model_year
3. Select the first 7 values
4. Get the underlying NumPy array. Let's call it X.
5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
6. Invert XTX.
7. Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
What's the sum of all the elements of the result?

In [30]:
car_df = df[df.origin == 'Asia'].copy()
car_df.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
8                   250            1.0       174.0     2714.219310   
12                  320            5.0       145.0     2783.868974   
14                  200            6.0       160.0     3582.687368   
20                  150            3.0       197.0     2231.808142   
21                  160            4.0       133.0     2659.431451   

    acceleration  model_year origin fuel_type         drivetrain  num_doors  \
8           10.3        2016   Asia    Diesel  Front-wheel drive       -1.0   
12          15.1        2010   Asia    Diesel    All-wheel drive        1.0   
14          14.9        2007   Asia    Diesel    All-wheel drive        0.0   
20          18.7        2011   Asia  Gasoline  Front-wheel drive        1.0   
21           NaN        2016   Asia  Gasoline  Front-wheel drive       -1.0   

    fuel_efficiency_mpg  
8             16.823554  
12            16.175820  
14            11.871091  
20            18.889083  
21            16.077730

In [35]:
car_df = car_df[['vehicle_weight', 'model_year']]
car_df = car_df.reset_index(drop = True)
car_df.head(7)

vehicle_weight  model_year
0     2714.219310        2016
1     2783.868974        2010
2     3582.687368        2007
3     2231.808142        2011
4     2659.431451        2016
5     2844.227534        2014
6     3761.994038        2019

In [38]:
X = car_df.head(7).values
XTX = X.T.dot(X)
XTX_inv = np.linalg.inv(XTX)
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])
w = (XTX_inv @ X.T) @ y
w

array([0.01386421, 0.5049067 ])

In [39]:
sum_of_w = w.sum()
print("The sum of all elements in w is:", sum_of_w)

The sum of all elements in w is: 0.5187709081074016
